File: server/src/Controllers/CarController.cs
using Server.Models;
using Microsoft.AspNetCore.Mvc;

namespace Server.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class CarController : ControllerBase
    {
        private readonly ICarService _carService;
        private readonly ILogger<AuthController> _logger;
        public CarController(ICarService carService, ILogger<AuthController> logger)
        {
            _carService = carService;
            _logger = logger;

        }

        private async Task<IActionResult> ExecuteAction(Func<Task<IActionResult>> action)
        {
            try
            {
                return await action();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "An error occurred while executing the action.");
                return StatusCode(500, new { message = "Internal Server Error" });
            }
        }

        // Add car
        [HttpPost("add")]
        public async Task<IActionResult> Add([FromBody] Car car)
        {
            return await ExecuteAction(async () =>
            {
                if (!ModelState.IsValid)
                {
                    return BadRequest(new { message = "Invalid data", errors = ModelState.Values.SelectMany(v => v.Errors).Select(e => e.ErrorMessage) });
                }

                var result = await _carService.Add(car);
                if (result == false)
                    return BadRequest(new { message = "Something Error" });
                return Ok(new { message = "Add Successfully" });
            });
        }

        // Delete car
        [HttpDelete("remove/{carId}")]
        public async Task<IActionResult> Delete([FromRoute] int carId)
        {
            return await ExecuteAction(async () =>
            {
                var result = await _carService.Delete(carId);
                if (result == false)
                    return NotFound(new { message = "Not found Car" });
                return Ok(new { message = "Delete Successfully" });
            });
        }

        // Update car
        [HttpPut("update/{carId}")]
        public async Task<IActionResult> Update([FromBody] Car car, [FromRoute] int carId)
        {
            return await ExecuteAction(async () =>
            {
                var result = await _carService.Update(car, carId);
                if (result == false)
                    return BadRequest(new { message = "Not found Car" });
                return Ok(new { message = "Update Successfully" });
            });
        }

        // Get all cars
        [HttpGet]
        public async Task<IActionResult> GetAll([FromQuery] int page)
        {
            return await ExecuteAction(async () =>
            {
                var result = await _carService.GetAll(page);
                if (result == null)
                    return NotFound(new { message = "Not found Cars" });
                return Ok(new { message = "Get Cars Successfully" , Car = result });
            });
        }

        // Get car By id
        [HttpGet("{id}")]
        public async Task<IActionResult> GetById([FromRoute] int id)
        {
            return await ExecuteAction(async () =>
            {
                var result = await _carService.GetById(id);
                if (result == null)
                    return NotFound(new { message = "Not found Car" });
                return Ok(new { message = "Get Car Successfully" , Car = result });
            });
        }

        // Download Export Certificate 
        [HttpGet("download/{id}")]
        public async Task<IActionResult> DownloadCertificate([FromRoute] int id)
        {
            var pdfStream = await _carService.GenerateCertificate(id);
            if (pdfStream == null || pdfStream.Length == 0)
                return BadRequest(new { message = "Error generating PDF" });
            return File(pdfStream.ToArray(), "application/pdf", $"Certificate_{id}.pdf");
        }


        // Get count
        [HttpGet("count")]
        public async Task<IActionResult> Count()
        {
            return await ExecuteAction(async () =>
            {
                var result = await _carService.Count();
                if (result == 0)
                    return NotFound(new { message = "Not found Car" });
                return Ok(new { message = "Get count Successfully", Car = result });
            });
        }

    }
}

File: server/src/controllers/car.controller.ts
import { Request, Response } from "express";
import CarService from "../services/car.service";
const { validationResult } = require("express-validator");

class CarController {
  private serviceInstance: CarService;
  private static instance: CarController;

  private constructor() {
    this.serviceInstance = CarService.getServiceInstance();
  }

  static getControllerInstance(): CarController {
    if (!CarController.instance) {
      CarController.instance = new CarController();
    }
    return CarController.instance;
  }

  private handleError(
    res: Response,
    message: string,
    error: unknown,
    status = 500
  ): Response {
    return res.status(status).send({
      message,
      error: error instanceof Error ? error.message : error,
    });
  }

  async addCar(req: Request, res: Response): Promise<Response> {
    try {
      const result = await this.serviceInstance.addCar(req.body);
      if (!result.success) return res.status(400).json(result);
      return res.status(201).json(result);
    } catch (error) {
      return this.handleError(res, "Failed to add car", error);
    }
  }

  async deleteCar(req: Request, res: Response): Promise<Response> {
    try {
      const result = await this.serviceInstance.deleteCar(req.params.carId);
      if (!result.success) return res.status(404).json(result);
      return res.status(200).json(result);
    } catch (error) {
      return this.handleError(res, "Failed to delete car", error);
    }
  }

  async updateCar(req: Request, res: Response): Promise<Response> {
    try {
      const result = await this.serviceInstance.updateCar(
        req.params.carId,
        req.body
      );
      if (!result.success) return res.status(400).json(result);
      return res.status(200).json(result);
    } catch (error) {
      return this.handleError(res, "Failed to update car", error);
    }
  }

  async getAllCars(req: Request, res: Response): Promise<Response> {
    try {
      const result = await this.serviceInstance.getAllCars(req.query);
      const count = await this.serviceInstance.getCarCount();
      if (!result.success) return res.status(404).json(result);
      return res.status(200).json({
        message: "Cars retrieved successfully",
        data: result,
        count: count,
      });
    } catch (error) {
      return this.handleError(res, "Failed to retrieve cars", error);
    }
  }

  async getCarById(req: Request, res: Response): Promise<Response> {
    try {
      const result = await this.serviceInstance.getCarById(req.params.carId);
      if (!result.success) return res.status(404).json(result);
      return res.status(200).json(result);
    } catch (error) {
      return this.handleError(res, "Failed to retrieve car", error);
    }
  }

  async getCarCount(req: Request, res: Response): Promise<Response> {
    try {
      const count = await this.serviceInstance.getCarCount();
      return res
        .status(200)
        .json({ message: "Car count retrieved successfully", count });
    } catch (error) {
      return this.handleError(res, "Failed to retrieve car count", error);
    }
  }

  async downloadCertificate(req: Request, res: Response): Promise<Response> {
    try {
      const pdfStream = await this.serviceInstance.generateCertificate(
        req.body
      );
      if (!pdfStream || pdfStream.length === 0) {
        return res.status(400).json({ message: "Error generating PDF" });
      }
      res.setHeader("Content-Type", "application/pdf");
      res.setHeader(
        "Content-Disposition",
        `attachment; filename="Certificate_${new Date().getTime()}.pdf"`
      );
      return res.send(pdfStream);
    } catch (error) {
      return this.handleError(
        res,
        "Failed to download please try again",
        error
      );
    }
  }

  async ViewCertificate(req: Request, res: Response): Promise<Response> {
    try {
      const pdfStream = await this.serviceInstance.generateCertificate(
        req.user
      );
      if (!pdfStream || pdfStream.length === 0) {
        return res.status(400).json({ message: "Error view PDF" });
      }
      res.setHeader("Content-Type", "application/pdf");
      res.setHeader(
        "Content-Disposition",
        `inline; filename="Certificate_${new Date().getTime()}.pdf"`
      );
      return res.send(pdfStream);
    } catch (error) {
      return this.handleError(
        res,
        `Failed to view please try again ${error}`,
        error
      );
    }
  }
}

export default CarController;
